### PR TITLE
moving DataFormats/RecoCandidate/interface/TrackAssociation.h into SimDataFormats

### PR DESCRIPTION
**Pull request by davidlange6**
reduces the number of incorrect dependencies on SimDataFormats within DataFormats packages.

### CODE REVIEW

**Comment by davidlange6** [comment]
please test

**Comment by davidlange6** [comment]
is DQMOffline supposed to need MC information?

**Comment by davidlange6** [comment]
please test

**Comment by jfernan2** [comment]
> is DQMOffline supposed to need MC information?

BTV might be using MC info in some old algos for efficiency calculations.
@XavierAtCERN @marco-link  can you please have a look? Is this still needed/used?

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/33186	bf1efcb 
- technical
- jenkins tests pass and comparisons with the baseline show no (relevant) differences

**Comment by jfernan2** [comment]
+1

**Comment by XavierAtCERN** [comment]
> > is DQMOffline supposed to need MC information?
> 
> BTV might be using MC info in some old algos for efficiency calculations.
> @XavierAtCERN @marco-link can you please have a look? Is this still needed/used?

Thanks for checking with us. :) Indeed, BTV has been historically using MC information to produce ROC curves based on the true flavour of the jets in offline DQM when running on MC. Not only true for old algorithms but for more recent ones as well. We are considering adding some specific selections (dileptonic ttbar, ...) which will enrich our monitoring plots in b-jets without the need for true MC information so we can discuss about removing ROC curves if needed. However, if it is not a major problem, we would be happy to keep them for completeness.

**Comment by davidlange6** [comment]
The pieces that need mc truth information should just move to Validation

**Comment by civanch** [comment]
+1

**Comment by XavierAtCERN** [comment]
> The pieces that need mc truth information should just move to Validation
> […](#)

Duly noted. Thanks for the feedback, we will work on that.

**Comment by silviodonato** [comment]
@cms-sw/alca-l2  do you have any objections?

**Comment by yuanchao** [comment]
According to the SDT test result, there are lost dict definitions:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a25045/13560/dupDict/lostDefs.txt

**Comment by davidlange6** [comment]
please test

**Comment by smuzaffar** [comment]
please test
newer IB is now available

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> I found compilation error when building:

apparently this is related to this PR

**Comment by davidlange6** [comment]
I don't reproduce the problem. @slava77 - do you have a specific change that you think its related to?

**Comment by davidlange6** [comment]
ok, I see why I don't reproduce it - cms-merge-topic doesn't check out RecoHI/HiTracking and something in the PR tests does (and RecoHI/HiTracking is missing basically all of its dependencies)

**Comment by davidlange6** [comment]
please test

**Comment by davidlange6** [comment]
@smuzaffar - how do I find  /cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33186/13874/install.sh?

***@***.*** ~]$ ls /cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33186
ls: cannot access /cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33186: No such file or directory

**Comment by smuzaffar** [comment]
@davidlange6 , looks like bot is not deploy it  ( https://github.com/cms-sw/cms-bot/blob/master/pr_testing/test_multiple_prs.sh#L903 ) if there are build errors. I can change it though

**Comment by davidlange6** [comment]
I think that would be great unless there is some significant overhead with that change - but at least the PR messages should be consistent with what is then available to the developer..

So, then how does one get the list of packages checked out by a PR test? It seems that something more than cms-merge-topic is being done?  

Eg, this continues to work..
```
  1002	13:15	scram p CMSSW_11_3_X_2021-03-30-2300
  1003	13:15	cd CMSSW_11_3_X_2021-03-30-2300
  1004	13:15	cmsenv
  1005	13:15	cd src
  1006	13:15	git cms-merge-topic 33186
  1007	13:16	scram b -j 8 
```

**Comment by smuzaffar** [comment]
try `git cms-checkdeps -a -A`. looks like `cms-merge-topic` only runs `git cms-checkdeps -a` (i.e. does not look for BuildFile changes)

**Comment by davidlange6** [comment]
Ok, thanks  again - that sounds like a bug in cms-checkdeps...

**Comment by smuzaffar** [comment]
there is `-A` option for `cms-merge-topic`  which instruct it to run cms-checkdeps with `-A`

**Comment by smuzaffar** [comment]
we can change merge-topic and checkdeps to always use `-A`

**Comment by davidlange6** [comment]
So -a is default and -A is optional - I’d propose we be consistent between the two (-A is basically always the right thing to do I guess)

**Comment by slava77** [comment]
@cmsbuild please test

to pick up cms-sw/cms-git-tools#118 / cms-sw/cmsdist#6779

**Comment by slava77** [comment]
abort test

perhaps it's better to wait for 03-31-2300 IB

**Comment by davidlange6** [comment]
please test

**Comment by slava77** [comment]
the header consistency error is `SimDataFormats/GeneratorProducts/interface/HepMCProduct.h:11:10: fatal error: HepMC/GenEvent.h: No such file or directory`

I'm not sure what's missing in this case.

**Comment by slava77** [comment]
the header consistency error is `SimDataFormats/GeneratorProducts/interface/HepMCProduct.h:11:10: fatal error: HepMC/GenEvent.h: No such file or directory`

I'm not sure what's missing in this case.


> the header consistency error is `SimDataFormats/GeneratorProducts/interface/HepMCProduct.h:11:10: fatal error: HepMC/GenEvent.h: No such file or directory`
> 
> I'm not sure what's missing in this case.

ah, it's still CMSSW_11_3_X_2021-03-31-2300

**Comment by slava77** [comment]
@cmsbuild please test

the last test was started about 2 hours too early before CMSSW_11_3_X_2021-04-01-1100 showed up

File: DataFormats/RecoCandidate/BuildFile.xml [hunk 1]
```diff
@@ -6,7 +6,6 @@
 <use name="DataFormats/EgammaReco"/>
 <use name="DataFormats/GsfTrackReco"/>
 <use name="clhep"/>
-<use name="SimDataFormats/TrackingAnalysis"/>
```
Comment (slava77) [inline DataFormats/RecoCandidate/BuildFile.xml:9]: apparently CommonTools/RecoAlgos/interface/CosmicTrackingParticleSelector.h (and CommonTools/RecoAlgos/BuildFile.xml ) depended on this to pull in the dependence on `SimDataFormats/TrackingAnalysis` .

So, we need to get it added to CommonTools/RecoAlgos/BuildFile.xml

**Comment by slava77** [comment]
ping

CommonTools/RecoAlgos/BuildFile.xml needs an update to add ` <use name="SimDataFormats/TrackingAnalysis"/> `

**Comment by davidlange6** [comment]
I should get back to this tomorrow.

**Comment by davidlange6** [comment]
please test

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/33186	64e8872
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no (relevant) differences; the warnings about duplicate dictionary definitions are apparently due to (temporary) double-counting what's available in the baseline release and in the local build

**Comment by yuanchao** [comment]
+1
* include "TrackAssociation.h" in both "MuonAssociatorEDProducer.h" and ".cc" and the later ".h" is also included in ".cc". Though seems no harm here.
* duplicated dictionary def. as mentioned by Slava
* clang tidy checks and statics build checks warnings are from other packages
* matrix tests and unit tests all passed

**Comment by christopheralanwest** [comment]
@davidlange6 This PR needs to be rebased to resolve conflicts.

**Comment by davidlange6** [comment]
please test

**Comment by slava77** [comment]
@cmsbuild please test

to get cleaner tests now that CMSSW_12_0_X_2021-04-16-1100 is available

**Comment by davidlange6** [comment]
please test

**Comment by davidlange6** [comment]
the error is unrelated to this PR. Any future comments?

**Comment by qliphy** [comment]
> the error is unrelated to this PR. Any future comments?

This should be fixed now. https://github.com/cms-sw/cmssw/issues/33486

**Comment by qliphy** [comment]
please test

**Comment by cecilecaillol** [comment]
+l1

**Comment by slava77** [comment]
+reconstruction 

for https://github.com/cms-sw/cmssw/pull/33186	c75d0d0

**Comment by christopheralanwest** [comment]
+alca

**Comment by jfernan2** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by silviodonato** [comment]
merge

**Comment by silviodonato** [comment]
@cmsbuild ping

**Comment by santocch** [comment]
+1

**Pull request merged**